File: templates/typescript-tests/src/ingest/engineTests.ts
import {
  OlapTable,
  ClickHouseEngines,
  Key,
  DateTime,
  Int8,
  ClickHouseTTL,
  ClickHouseDefault,
  ClickHouseCodec,
  UInt32,
} from "@514labs/moose-lib";

/**
 * Test interfaces for various engine configurations
 * These tables test all supported ClickHouse engines to ensure
 * they can be properly created and configured.
 */

// Test data model for engine testing
export interface EngineTestData {
  id: Key<string>;
  timestamp: DateTime;
  value: number;
  category: string;
  version: number;
  isDeleted: boolean; // For ReplacingMergeTree soft deletes (UInt8 in ClickHouse)
}

// Test data model for CollapsingMergeTree and VersionedCollapsingMergeTree testing
export interface CollapsingTestData extends EngineTestData {
  sign: Int8; // For CollapsingMergeTree (1 = state, -1 = cancel)
}

export interface EngineTestDataSample {
  id: string;
  timestamp: DateTime;
  value: number;
  category: string;
  version: number;
  isDeleted: boolean; // For ReplacingMergeTree soft deletes (UInt8 in ClickHouse)
}

// Table with TTL: delete rows older than 90 days, delete email after 30 days
export interface TTLTestData {
  id: Key<string>;
  timestamp: DateTime;
  email: string & ClickHouseTTL<"timestamp + INTERVAL 30 DAY">;
}

export const TTLTable = new OlapTable<TTLTestData>("TTLTable", {
  engine: ClickHouseEngines.MergeTree,
  orderByFields: ["id", "timestamp"],
  ttl: "timestamp + INTERVAL 90 DAY DELETE",
});

// Table with DEFAULT values for testing DEFAULT removal
export interface DefaultTestData {
  id: Key<string>;
  timestamp: DateTime;
  status: string & ClickHouseDefault<"'pending'">;
  count: UInt32 & ClickHouseDefault<"0">;
}

export const DefaultTable = new OlapTable<DefaultTestData>("DefaultTable", {
  engine: ClickHouseEngines.MergeTree,
  orderByFields: ["id", "timestamp"],
});

// Test MergeTree engine (default)
export const MergeTreeTable = new OlapTable<EngineTestData>("MergeTreeTest", {
  engine: ClickHouseEngines.MergeTree,
  orderByFields: ["id", "timestamp"],
});

// Test MergeTree with orderByExpression (equivalent to fields)
export const MergeTreeTableExpr = new OlapTable<EngineTestData>(
  "MergeTreeTestExpr",
  {
    engine: ClickHouseEngines.MergeTree,
    orderByExpression: "(id, timestamp)",
  },
);

// Test ReplacingMergeTree engine with basic deduplication
export const ReplacingMergeTreeBasicTable = new OlapTable<EngineTestData>(
  "ReplacingMergeTreeBasic",
  {
    engine: ClickHouseEngines.ReplacingMergeTree,
    orderByFields: ["id"],
  },
);

// Test ReplacingMergeTree engine with version column
export const ReplacingMergeTreeVersionTable = new OlapTable<EngineTestData>(
  "ReplacingMergeTreeVersion",
  {
    engine: ClickHouseEngines.ReplacingMergeTree,
    orderByFields: ["id"],
    ver: "version",
  },
);

// Test ReplacingMergeTree engine with version and soft delete
export const ReplacingMergeTreeSoftDeleteTable = new OlapTable<EngineTestData>(
  "ReplacingMergeTreeSoftDelete",
  {
    engine: ClickHouseEngines.ReplacingMergeTree,
    orderByFields: ["id"],
    ver: "version",
    isDeleted: "isDeleted",
  },
);

// Test SummingMergeTree engine
export const SummingMergeTreeTable = new OlapTable<EngineTestData>(
  "SummingMergeTreeTest",
  {
    engine: ClickHouseEngines.SummingMergeTree,
    orderByFields: ["id", "category"],
  },
);

// Test AggregatingMergeTree engine
export const AggregatingMergeTreeTable = new OlapTable<EngineTestData>(
  "AggregatingMergeTreeTest",
  {
    engine: ClickHouseEngines.AggregatingMergeTree,
    orderByFields: ["id", "category"],
  },
);

// Test CollapsingMergeTree engine
export const CollapsingMergeTreeTable = new OlapTable<CollapsingTestData>(
  "CollapsingMergeTreeTest",
  {
    engine: ClickHouseEngines.CollapsingMergeTree,
    sign: "sign",
    orderByFields: ["id", "timestamp"],
  },
);

// Test VersionedCollapsingMergeTree engine
export const VersionedCollapsingMergeTreeTable =
  new OlapTable<CollapsingTestData>("VersionedCollapsingMergeTreeTest", {
    engine: ClickHouseEngines.VersionedCollapsingMergeTree,
    sign: "sign",
    ver: "version",
    orderByFields: ["id", "timestamp"],
  });

// Test SummingMergeTree engine with columns
export const SummingMergeTreeWithColumnsTable = new OlapTable<EngineTestData>(
  "SummingMergeTreeWithColumnsTest",
  {
    engine: ClickHouseEngines.SummingMergeTree,
    orderByFields: ["id", "category"],
    columns: ["value"],
  },
);

// Test ReplicatedMergeTree engine (with explicit keeper params - for self-hosted)
export const ReplicatedMergeTreeTable = new OlapTable<EngineTestData>(
  "ReplicatedMergeTreeTest",
  {
    engine: ClickHouseEngines.ReplicatedMergeTree,
    keeperPath:
      "/clickhouse/tables/{database}/{shard}/replicated_merge_tree_test",
    replicaName: "{replica}",
    orderByFields: ["id", "timestamp"],
  },
);

// Test ReplicatedReplacingMergeTree engine with version column
export const ReplicatedReplacingMergeTreeTable = new OlapTable<EngineTestData>(
  "ReplicatedReplacingMergeTreeTest",
  {
    engine: ClickHouseEngines.ReplicatedReplacingMergeTree,
    keeperPath:
      "/clickhouse/tables/{database}/{shard}/replicated_replacing_test",
    replicaName: "{replica}",
    ver: "version",
    orderByFields: ["id"],
  },
);

// Test ReplicatedReplacingMergeTree with soft delete
export const ReplicatedReplacingSoftDeleteTable = new OlapTable<EngineTestData>(
  "ReplicatedReplacingSoftDeleteTest",
  {
    engine: ClickHouseEngines.ReplicatedReplacingMergeTree,
    keeperPath:
      "/clickhouse/tables/{database}/{shard}/replicated_replacing_sd_test",
    replicaName: "{replica}",
    ver: "version",
    isDeleted: "isDeleted",
    orderByFields: ["id"],
  },
);

// Test ReplicatedAggregatingMergeTree engine
export const ReplicatedAggregatingMergeTreeTable =
  new OlapTable<EngineTestData>("ReplicatedAggregatingMergeTreeTest", {
    engine: ClickHouseEngines.ReplicatedAggregatingMergeTree,
    keeperPath:
      "/clickhouse/tables/{database}/{shard}/replicated_aggregating_test",
    replicaName: "{replica}",
    orderByFields: ["id", "category"],
  });

// Test ReplicatedSummingMergeTree engine
export const ReplicatedSummingMergeTreeTable = new OlapTable<EngineTestData>(
  "ReplicatedSummingMergeTreeTest",
  {
    engine: ClickHouseEngines.ReplicatedSummingMergeTree,
    keeperPath: "/clickhouse/tables/{database}/{shard}/replicated_summing_test",
    replicaName: "{replica}",
    columns: ["value"],
    orderByFields: ["id", "category"],
  },
);

// Test ReplicatedCollapsingMergeTree engine
export const ReplicatedCollapsingMergeTreeTable =
  new OlapTable<CollapsingTestData>("ReplicatedCollapsingMergeTreeTest", {
    engine: ClickHouseEngines.ReplicatedCollapsingMergeTree,
    keeperPath:
      "/clickhouse/tables/{database}/{shard}/replicated_collapsing_test",
    replicaName: "{replica}",
    sign: "sign",
    orderByFields: ["id", "timestamp"],
  });

// Test ReplicatedVersionedCollapsingMergeTree engine
export const ReplicatedVersionedCollapsingMergeTreeTable =
  new OlapTable<CollapsingTestData>(
    "ReplicatedVersionedCollapsingMergeTreeTest",
    {
      engine: ClickHouseEngines.ReplicatedVersionedCollapsingMergeTree,
      keeperPath:
        "/clickhouse/tables/{database}/{shard}/replicated_versioned_collapsing_test",
      replicaName: "{replica}",
      sign: "sign",
      ver: "version",
      orderByFields: ["id", "timestamp"],
    },
  );

// Test SAMPLE BY clause for data sampling
export const SampleByTable = new OlapTable<EngineTestDataSample>(
  "SampleByTest",
  {
    engine: ClickHouseEngines.MergeTree,
    orderByExpression: "cityHash64(id)",
    sampleByExpression: "cityHash64(id)",
  },
);

// Table for testing MODIFY COLUMN with comment + codec combinations
export interface CommentCodecTestData {
  id: Key<string>;
  timestamp: DateTime;
  /** Raw data payload */
  data: string & ClickHouseCodec<"ZSTD(3)">;
  /** Measurement value */
  metric: number & ClickHouseCodec<"ZSTD(1)">;
  /** Classification label */
  label: string;
  compressed: string & ClickHouseCodec<"LZ4">;
}

export const CommentCodecTable = new OlapTable<CommentCodecTestData>(
  "CommentCodecTest",
  {
    engine: ClickHouseEngines.MergeTree,
    orderByFields: ["id", "timestamp"],
  },
);

// Note: S3Queue engine testing is more complex as it requires S3 configuration
// and external dependencies, so it's not included in this basic engine test suite.
// For S3Queue testing, see the dedicated S3 integration tests.

// Test Buffer engine - buffers writes before flushing to destination table
// First create the destination table
export const BufferDestinationTable = new OlapTable<EngineTestData>(
  "BufferDestinationTest",
  {
    engine: ClickHouseEngines.MergeTree,
    orderByFields: ["id", "timestamp"],
  },
);

// Then create the buffer table that points to it
export const BufferTable = new OlapTable<EngineTestData>("BufferTest", {
  engine: ClickHouseEngines.Buffer,
  targetDatabase: "local",
  targetTable: "BufferDestinationTest",
  numLayers: 16,
  minTime: 10,
  maxTime: 100,
  minRows: 10000,
  maxRows: 1000000,
  minBytes: 10485760,
  maxBytes: 104857600,
});

// Test Merge engine - virtual read-only view over tables matching a regex
// Source tables that the Merge table will read from
export const MergeSourceA = new OlapTable<EngineTestData>("MergeSourceA", {
  engine: ClickHouseEngines.MergeTree,
  orderByFields: ["id", "timestamp"],
});

export const MergeSourceB = new OlapTable<EngineTestData>("MergeSourceB", {
  engine: ClickHouseEngines.MergeTree,
  orderByFields: ["id", "timestamp"],
});

// Merge table: reads from all tables matching ^MergeSource.* in the current database
export const MergeTable = new OlapTable<EngineTestData>("MergeTest", {
  engine: ClickHouseEngines.Merge,
  sourceDatabase: "currentDatabase()",
  tablesRegexp: "^MergeSource.*",
});

/**
 * Export all test tables for verification that engine configurations
 * can be properly instantiated and don't throw errors during table creation.
 */
export const allEngineTestTables = [
  MergeTreeTable,
  MergeTreeTableExpr,
  ReplacingMergeTreeBasicTable,
  ReplacingMergeTreeVersionTable,
  ReplacingMergeTreeSoftDeleteTable,
  SummingMergeTreeTable,
  SummingMergeTreeWithColumnsTable,
  AggregatingMergeTreeTable,
  CollapsingMergeTreeTable,
  VersionedCollapsingMergeTreeTable,
  ReplicatedMergeTreeTable,
  ReplicatedReplacingMergeTreeTable,
  ReplicatedReplacingSoftDeleteTable,
  ReplicatedAggregatingMergeTreeTable,
  ReplicatedSummingMergeTreeTable,
  ReplicatedCollapsingMergeTreeTable,
  ReplicatedVersionedCollapsingMergeTreeTable,
  SampleByTable,
  TTLTable,
  MergeSourceA,
  MergeSourceB,
  MergeTable,
  BufferDestinationTable,
  BufferTable,
  DefaultTable,
  CommentCodecTable,
];

File: templates/python-tests/src/ingest/engine_tests.py
# Test all supported ClickHouse engines to ensure proper configuration
# These tables verify that all engine types can be created and configured correctly

from moose_lib import (
    OlapTable,
    OlapConfig,
    Key,
    Int8,
    ClickHouseTTL,
    ClickHouseCodec,
    clickhouse_default,
    FixedString,
)
from moose_lib.blocks import (
    MergeTreeEngine,
    ReplacingMergeTreeEngine,
    SummingMergeTreeEngine,
    AggregatingMergeTreeEngine,
    CollapsingMergeTreeEngine,
    VersionedCollapsingMergeTreeEngine,
    ReplicatedMergeTreeEngine,
    ReplicatedReplacingMergeTreeEngine,
    ReplicatedAggregatingMergeTreeEngine,
    ReplicatedSummingMergeTreeEngine,
    ReplicatedCollapsingMergeTreeEngine,
    ReplicatedVersionedCollapsingMergeTreeEngine,
    BufferEngine,
    MergeEngine,
    # S3QueueEngine - requires S3 configuration, tested separately
)
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional, Annotated, List


class EngineTestData(BaseModel):
    """Test data model for engine testing"""

    id: Key[str]
    timestamp: datetime
    value: int
    category: str
    version: int
    is_deleted: bool  # For ReplacingMergeTree soft deletes (UInt8 in ClickHouse)


# Table with TTL: delete rows older than 90 days, delete email after 30 days
class TTLTestData(BaseModel):
    id: Key[str]
    timestamp: datetime
    email: Annotated[str, ClickHouseTTL("timestamp + INTERVAL 30 DAY")]


ttl_table = OlapTable[TTLTestData](
    "TTLTable",
    OlapConfig(
        engine=MergeTreeEngine(),
        order_by_fields=["id", "timestamp"],
        ttl="timestamp + INTERVAL 90 DAY DELETE",
    ),
)


# Table with DEFAULT values for testing DEFAULT removal
class DefaultTestData(BaseModel):
    id: Key[str]
    timestamp: datetime
    status: Annotated[str, clickhouse_default("'pending'")]
    count: Annotated[int, clickhouse_default("0"), "uint32"]


default_table = OlapTable[DefaultTestData](
    "DefaultTable",
    OlapConfig(
        engine=MergeTreeEngine(),
        order_by_fields=["id", "timestamp"],
    ),
)

# Table with FixedString: test direct usage and type aliases
# Type alias for MAC address (17 bytes for format "XX:XX:XX:XX:XX:XX")
type MacAddress = Annotated[str, FixedString(17)]


class FixedStringTestData(BaseModel):
    id: Key[str]
    timestamp: datetime
    # Direct FixedString usage
    md5_hash: Annotated[str, FixedString(16)]  # 16-byte MD5
    ipv6_address: Annotated[str, FixedString(16)]  # 16-byte IPv6
    # Type alias usage
    mac_address: MacAddress  # Should become FixedString(17)
    mac_addresses: List[MacAddress]  # Should become Array(FixedString(17))


fixedstring_table = OlapTable[FixedStringTestData](
    "FixedStringTest",
    OlapConfig(
        engine=MergeTreeEngine(),
        order_by_fields=["id", "timestamp"],
    ),
)


class CollapsingTestData(EngineTestData):
    """Test data model for CollapsingMergeTree and VersionedCollapsingMergeTree testing"""

    sign: Int8  # For CollapsingMergeTree (1 = state, -1 = cancel)


class EngineTestDataSample(BaseModel):
    """Test data model for engine testing"""

    id: str
    timestamp: datetime
    value: int
    category: str
    version: int
    is_deleted: bool  # For ReplacingMergeTree soft deletes (UInt8 in ClickHouse)


# Test MergeTree engine (default)
merge_tree_table = OlapTable[EngineTestData](
    "MergeTreeTest",
    OlapConfig(engine=MergeTreeEngine(), order_by_fields=["id", "timestamp"]),
)

# Test MergeTree with order_by_expression (equivalent to fields)
merge_tree_table_expr = OlapTable[EngineTestData](
    "MergeTreeTestExpr",
    OlapConfig(
        engine=MergeTreeEngine(),
        order_by_expression="(id, timestamp)",
    ),
)

# Test ReplacingMergeTree engine with basic deduplication
replacing_merge_tree_basic_table = OlapTable[EngineTestData](
    "ReplacingMergeTreeBasic",
    OlapConfig(engine=ReplacingMergeTreeEngine(), order_by_fields=["id"]),
)

# Test ReplacingMergeTree engine with version column
replacing_merge_tree_version_table = OlapTable[EngineTestData](
    "ReplacingMergeTreeVersion",
    OlapConfig(engine=ReplacingMergeTreeEngine(ver="version"), order_by_fields=["id"]),
)

# Test ReplacingMergeTree engine with version and soft delete
replacing_merge_tree_soft_delete_table = OlapTable[EngineTestData](
    "ReplacingMergeTreeSoftDelete",
    OlapConfig(
        engine=ReplacingMergeTreeEngine(ver="version", is_deleted="is_deleted"),
        order_by_fields=["id"],
    ),
)

# Test SummingMergeTree engine
summing_merge_tree_table = OlapTable[EngineTestData](
    "SummingMergeTreeTest",
    OlapConfig(engine=SummingMergeTreeEngine(), order_by_fields=["id", "category"]),
)

# Test AggregatingMergeTree engine
aggregating_merge_tree_table = OlapTable[EngineTestData](
    "AggregatingMergeTreeTest",
    OlapConfig(engine=AggregatingMergeTreeEngine(), order_by_fields=["id", "category"]),
)

# Test CollapsingMergeTree engine
collapsing_merge_tree_table = OlapTable[CollapsingTestData](
    "CollapsingMergeTreeTest",
    OlapConfig(
        engine=CollapsingMergeTreeEngine(sign="sign"),
        order_by_fields=["id", "timestamp"],
    ),
)

# Test VersionedCollapsingMergeTree engine
versioned_collapsing_merge_tree_table = OlapTable[CollapsingTestData](
    "VersionedCollapsingMergeTreeTest",
    OlapConfig(
        engine=VersionedCollapsingMergeTreeEngine(sign="sign", ver="version"),
        order_by_fields=["id", "timestamp"],
    ),
)

# Test ReplicatedMergeTree engine (with explicit keeper params - for self-hosted)
replicated_merge_tree_table = OlapTable[EngineTestData](
    "ReplicatedMergeTreeTest",
    OlapConfig(
        engine=ReplicatedMergeTreeEngine(
            keeper_path="/clickhouse/tables/{database}/{shard}/replicated_merge_tree_test",
            replica_name="{replica}",
        ),
        order_by_fields=["id", "timestamp"],
    ),
)

# Test ReplicatedMergeTree engine (Cloud-compatible - no keeper params)
# In dev mode, Moose automatically injects default parameters
# In production, ClickHouse uses its automatic configuration
replicated_merge_tree_cloud_table = OlapTable[EngineTestData](
    "ReplicatedMergeTreeCloudTest",
    OlapConfig(
        engine=ReplicatedMergeTreeEngine(),  # No params - uses server defaults (Cloud compatible)
        order_by_fields=["id", "timestamp"],
    ),
)

# Test ReplicatedReplacingMergeTree engine with version column
replicated_replacing_merge_tree_table = OlapTable[EngineTestData](
    "ReplicatedReplacingMergeTreeTest",
    OlapConfig(
        engine=ReplicatedReplacingMergeTreeEngine(
            keeper_path="/clickhouse/tables/{database}/{shard}/replicated_replacing_test",
            replica_name="{replica}",
            ver="version",
        ),
        order_by_fields=["id"],
    ),
)

# Test ReplicatedReplacingMergeTree with soft delete
replicated_replacing_soft_delete_table = OlapTable[EngineTestData](
    "ReplicatedReplacingSoftDeleteTest",
    OlapConfig(
        engine=ReplicatedReplacingMergeTreeEngine(
            keeper_path="/clickhouse/tables/{database}/{shard}/replicated_replacing_sd_test",
            replica_name="{replica}",
            ver="version",
            is_deleted="is_deleted",
        ),
        order_by_fields=["id"],
    ),
)

# Test ReplicatedAggregatingMergeTree engine
replicated_aggregating_merge_tree_table = OlapTable[EngineTestData](
    "ReplicatedAggregatingMergeTreeTest",
    OlapConfig(
        engine=ReplicatedAggregatingMergeTreeEngine(
            keeper_path="/clickhouse/tables/{database}/{shard}/replicated_aggregating_test",
            replica_name="{replica}",
        ),
        order_by_fields=["id", "category"],
    ),
)

# Test ReplicatedSummingMergeTree engine
replicated_summing_merge_tree_table = OlapTable[EngineTestData](
    "ReplicatedSummingMergeTreeTest",
    OlapConfig(
        engine=ReplicatedSummingMergeTreeEngine(
            keeper_path="/clickhouse/tables/{database}/{shard}/replicated_summing_test",
            replica_name="{replica}",
            columns=["value"],
        ),
        order_by_fields=["id", "category"],
    ),
)

# Test ReplicatedCollapsingMergeTree engine
replicated_collapsing_merge_tree_table = OlapTable[CollapsingTestData](
    "ReplicatedCollapsingMergeTreeTest",
    OlapConfig(
        engine=ReplicatedCollapsingMergeTreeEngine(
            keeper_path="/clickhouse/tables/{database}/{shard}/replicated_collapsing_test",
            replica_name="{replica}",
            sign="sign",
        ),
        order_by_fields=["id", "timestamp"],
    ),
)

# Test ReplicatedVersionedCollapsingMergeTree engine
replicated_versioned_collapsing_merge_tree_table = OlapTable[CollapsingTestData](
    "ReplicatedVersionedCollapsingMergeTreeTest",
    OlapConfig(
        engine=ReplicatedVersionedCollapsingMergeTreeEngine(
            keeper_path="/clickhouse/tables/{database}/{shard}/replicated_versioned_collapsing_test",
            replica_name="{replica}",
            sign="sign",
            ver="version",
        ),
        order_by_fields=["id", "timestamp"],
    ),
)

# Test SAMPLE BY clause for data sampling
sample_by_table = OlapTable[EngineTestDataSample](
    "SampleByTest",
    OlapConfig(
        engine=MergeTreeEngine(),
        order_by_expression="cityHash64(id)",
        sample_by_expression="cityHash64(id)",
    ),
)


# Table for testing MODIFY COLUMN with comment + codec combinations
class CommentCodecTestData(BaseModel):
    id: Key[str]
    timestamp: datetime
    data: Annotated[str, ClickHouseCodec("ZSTD(3)")] = Field(
        description="Raw data payload"
    )
    metric: Annotated[float, ClickHouseCodec("ZSTD(1)")] = Field(
        description="Measurement value"
    )
    label: str = Field(description="Classification label")
    compressed: Annotated[str, ClickHouseCodec("LZ4")]


comment_codec_table = OlapTable[CommentCodecTestData](
    "CommentCodecTest",
    OlapConfig(
        engine=MergeTreeEngine(),
        order_by_fields=["id", "timestamp"],
    ),
)

# Note: S3Queue engine testing is more complex as it requires S3 configuration
# and external dependencies, so it's not included in this basic engine test suite.
# For S3Queue testing, see the dedicated S3 integration tests.

# Test Merge engine - virtual read-only view over tables matching a regex
# Source tables that the Merge table will read from
merge_source_a = OlapTable[EngineTestData](
    "MergeSourceA",
    OlapConfig(engine=MergeTreeEngine(), order_by_fields=["id", "timestamp"]),
)

merge_source_b = OlapTable[EngineTestData](
    "MergeSourceB",
    OlapConfig(engine=MergeTreeEngine(), order_by_fields=["id", "timestamp"]),
)

# Merge table: reads from all tables matching ^MergeSource.* in the current database
merge_table = OlapTable[EngineTestData](
    "MergeTest",
    OlapConfig(
        engine=MergeEngine(
            source_database="currentDatabase()",
            tables_regexp="^MergeSource.*",
        )
    ),
)

# Test Buffer engine - buffers writes before flushing to destination table
# First create the destination table
buffer_destination_table = OlapTable[EngineTestData](
    "BufferDestinationTest",
    OlapConfig(engine=MergeTreeEngine(), order_by_fields=["id", "timestamp"]),
)

# Then create the buffer table that points to it
buffer_table = OlapTable[EngineTestData](
    "BufferTest",
    OlapConfig(
        engine=BufferEngine(
            target_database="local",
            target_table="BufferDestinationTest",
            num_layers=16,
            min_time=10,
            max_time=100,
            min_rows=10000,
            max_rows=1000000,
            min_bytes=10485760,
            max_bytes=104857600,
        ),
    ),
)

# Export all test tables for verification that engine configurations
# can be properly instantiated and don't throw errors during table creation
all_engine_test_tables = [
    merge_tree_table,
    merge_tree_table_expr,
    replacing_merge_tree_basic_table,
    replacing_merge_tree_version_table,
    replacing_merge_tree_soft_delete_table,
    summing_merge_tree_table,
    aggregating_merge_tree_table,
    collapsing_merge_tree_table,
    versioned_collapsing_merge_tree_table,
    replicated_merge_tree_table,
    replicated_merge_tree_cloud_table,
    replicated_replacing_merge_tree_table,
    replicated_replacing_soft_delete_table,
    replicated_aggregating_merge_tree_table,
    replicated_summing_merge_tree_table,
    replicated_collapsing_merge_tree_table,
    replicated_versioned_collapsing_merge_tree_table,
    sample_by_table,
    ttl_table,
    merge_source_a,
    merge_source_b,
    merge_table,
    buffer_destination_table,
    buffer_table,
    default_table,
    fixedstring_table,
    comment_codec_table,
]
